File: LeetCode/0777. Swap Adjacent in LR String/solution.cpp
/**
 * 94 / 94 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 7.2 MB 
 */
class Solution {
public:
    bool canTransform(string start, string end) {
        int n = start.size();
        int i = 0, j = 0;
        for (;;) {
            while (i < n && start[i] == 'X') ++ i;
            while (j < n && end[j] == 'X') ++ j;
            if (i == n && j == n) return true;
            if (i == n || j == n || start[i] != end[j]) return false;
            if (start[i] == 'R' && i > j) return false;
            if (start[i] == 'L' && i < j) return false;
            ++i;
            ++j;
        }
        return true;
    }
};

File: LeetCode/0777. Swap Adjacent in LR String/solution.py
"""
94 / 94 test cases passed.
Runtime: 48 ms
Memory Usage: 15 MB
"""
class Solution:
    def canTransform(self, start: str, end: str) -> bool:
        n, i, j = len(start), 0, 0
        while True:
            while i < n and start[i] == 'X':
                i += 1
            while j < n and end[j] == 'X':
                j += 1
            if i == n and j == n:
                return True
            if i == n or j == n or start[i] != end[j]:
                return False
            if start[i] == 'L' and i < j:
                return False
            if start[i] == 'R' and i > j:
                return False
            i += 1
            j += 1
        return True
